src/Workspaces/Remote/Core/Diagnostics/DiagnosticComputer.cs

```diff
@@ -124,28 +165,109 @@ private static ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(BidirectionalMap<
             return builder.ToImmutable();
         }
 
-        private static BidirectionalMap<string, DiagnosticAnalyzer> CreateAnalyzerMap(Project project)
+        private sealed class CompilationWithAnalyzersData
         {
-            // we could consider creating a service so that we don't do this repeatedly if this shows up as perf cost
-            using var pooledObject = SharedPools.Default<HashSet<object>>().GetPooledObject();
-            using var pooledMap = SharedPools.Default<Dictionary<string, DiagnosticAnalyzer>>().GetPooledObject();
-            var referenceSet = pooledObject.Object;
-            var analyzerMap = pooledMap.Object;
-
-            // this follow what we do in DiagnosticAnalyzerInfoCache.CheckAnalyzerReferenceIdentity
-            foreach (var reference in project.Solution.AnalyzerReferences.Concat(project.AnalyzerReferences))
+            /// <summary>
+            /// Cache of <see cref="CompilationWithAnalyzers"/> and a map from analyzer IDs to <see cref="DiagnosticAnalyzer"/>s
+            /// for all analyzers for each project.
+            /// The <see cref="CompilationWithAnalyzers"/> instance is shared between all the following document analyses modes for the project:
+            ///  1. Span-based analysis for active document (lightbulb)
+            ///  2. Background analysis for active and open documents.
+            ///  
+            /// NOTE: We do not re-use this cache for project analysis as it leads to significant memory increase in the OOP process,
+            /// and CWT does not seem to drop entries until ForceGC happens.
+            /// </summary>
+            private static readonly ConditionalWeakTable<Project, CompilationWithAnalyzersData> s_cache
+                = new ConditionalWeakTable<Project, CompilationWithAnalyzersData>();
+
+            private CompilationWithAnalyzersData(CompilationWithAnalyzers compilationWithAnalyzers, BidirectionalMap<string, DiagnosticAnalyzer> analyzerToIdMap)
+            {
+                CompilationWithAnalyzers = compilationWithAnalyzers;
+                AnalyzerToIdMap = analyzerToIdMap;
+            }
+
+            public BidirectionalMap<string, DiagnosticAnalyzer> AnalyzerToIdMap { get; }
+            public CompilationWithAnalyzers CompilationWithAnalyzers { get; }
+
```

Comment:
Nit: move these before the constructor, so that it's clear what comprises the state of the object.